src/Data/Tree/DUAL/Internal.hs

```diff
@@ -146,178 +119,137 @@ pullU t@(Annot _ (DUALTreeU (u, _))) = pack (u, t)
+--   the identity. O(1)
+annot :: a -> DUALTree d u a l -> DUALTree d u a l
+annot _ EmptyDUAL      = EmptyDUAL
+annot a (DUALTree u t) = DUALTree u (Annot a t)
 
 -- | Apply a @d@ annotation at the root of a tree, transforming all
 --   @u@ annotations by the action of @d@.
-applyD :: (Semigroup d, Semigroup u, Action d u)
-       => d -> DUALTree d u a l -> DUALTree d u a l
-applyD = act . DAct
-
--- | Decompose a DUAL-tree into either @Nothing@ (if empty) or a
---   top-level cached @u@ annotation paired with a non-empty
---   DUAL-tree.
-nonEmpty :: DUALTree d u a l -> Maybe (u, DUALTreeNE d u a l)
-nonEmpty = fmap unpack . getOption . unpack
-
--- | Get the @u@ annotation at the root, or @Nothing@ if the tree is
---   empty.
-getU :: DUALTree d u a l -> Maybe u
-getU = fmap fst . nonEmpty
+down :: (Semigroup d, Semigroup u, Action d u) => d -> DUALTree d u a l -> DUALTree d u a l
+down _ EmptyDUAL      = EmptyDUAL
+down d (DUALTree u t) = DUALTree (act d u) $ case t of
+  Down d' t' -> Down (d <> d') t'
+  _          -> Down d t
 
-------------------------------------------------------------
--- Maps
-------------------------------------------------------------
+-- | Get the up annotation of a non-empty DUALTree.
+getU :: DUALTree d u a l -> Maybe u
+getU (DUALTree u _) = Just u
+getU _              = Nothing
 
--- XXX todo: try adding Map as a constructor, so we can delay the
--- mapping until the end too?
-
--- | Map a function (which must be a monoid homomorphism, and commute
---   with the action of @d@) over all the @u@ annotations in a non-empty
---   DUAL-tree.
-mapUNE :: (u -> u') -> DUALTreeNE d u a l -> DUALTreeNE d u' a l
-mapUNE f (Leaf u l)  = Leaf (f u) l
-mapUNE f (LeafU u)   = LeafU (f u)
-mapUNE f (Concat ts) = Concat ((NEL.map . mapUU) f ts)
-mapUNE f (Act d t)   = Act d (mapUU f t)
-mapUNE f (Annot a t) = Annot a (mapUU f t)
-
--- | Map a function (which must be a monoid homomorphism, and commute
---   with the action of @d@) over all the @u@ annotations in a
---   non-empty DUAL-tree paired with its cached @u@ value.
-mapUU :: (u -> u') -> DUALTreeU d u a l -> DUALTreeU d u' a l
-mapUU f = over DUALTreeU (f *** mapUNE f)
-
--- | Map a function over all the @u@ annotations in a DUAL-tree.  The
---   function must be a monoid homomorphism, and must commute with the
---   action of @d@ on @u@.  That is, to use @mapU f@ safely it must be
---   the case that
---
---     * @f mempty == mempty@
---
---     * @f (u1 \<\> u2) == f u1 \<\> f u2@
---
---     * @f (act d u) == act d (f u)@
---
+-- | Map over the @u@ annotation of a DUALTree.
 mapU :: (u -> u') -> DUALTree d u a l -> DUALTree d u' a l
-mapU = over DUALTree . fmap . mapUU
+mapU f (DUALTree u t) = DUALTree (f u) t
+mapU _ _              = EmptyDUAL
+
+-- | Apply a @u@ annotation of a DUALTree on the left. Makes a 'leafU'
+--   for an empty tree.
+preapplyU :: Semigroup u => u -> DUALTree d u a l -> DUALTree d u a l
+preapplyU u (DUALTree u' t) = DUALTree (u' <> u) t
```

Comment:
The pre / post where from the [old names](https://github.com/diagrams/dual-tree/blob/master/src/Data/Tree/DUAL/Internal.hs#L187-L197): `applyUpre`, `applyUpost`, although you're right, it looks like I've managed to mix them up.